src/anndata/_core/anndata.py

```diff
@@ -640,10 +640,24 @@ def X(self, value: np.ndarray | sparse.spmatrix | SpArray | None):
             or (self.n_vars == 1 and self.n_obs == len(value))
             or (self.n_obs == 1 and self.n_vars == len(value))
         ):
-            if not np.isscalar(value) and self.shape != value.shape:
-                # For assigning vector of values to 2d array or matrix
-                # Not necessary for row of 2d array
-                value = value.reshape(self.shape)
+            if not np.isscalar(value):
+                if self.is_view and any(
+                    isinstance(idx, np.ndarray)
+                    and len(np.unique(idx)) != len(idx.ravel())
+                    for idx in [oidx, vidx]
+                ):
+                    warnings.warn(
+                        "You are attempting to set `X` to a matrix on a view which has non-unique indices. "
+                        "The resulting `adata.X` will likely not equal the value to which you set it. "
+                        "To avoid this potential issue, please make a copy of the data first. "
+                        "In the future, this operation will throw an error.",
```

Comment:
Why do you want to start with a warning now and not just immediately go to the error behavior? To give people that are currently doing this a bit of a grace period where their code still works? I think that this case is rare enough so that you can immediately jump into the error behavior, no?

(I don't have a strong opinion on this but I thought I'd ask anyways)

Reply:
Yes, I generally have a stricter definition of breaking change than others.